src/rewrite_infinite_loops.cpp

```diff
@@ -0,0 +1,310 @@
+/**
+ * @file rewrite_infinite_loops.cpp
+ *
+ * @author  Alex Henrie
+ * @license GPL v2+
+ */
+
+#include "rewrite_infinite_loops.h"
+
+#include "chunk.h"
+#include "uncrustify.h"
+
+
+using namespace uncrustify;
+
+
+static Chunk *find_end_brace(Chunk *pc)
+{
+   while (!chunk_is_opening_brace(pc))
+   {
+      pc = pc->GetNextNcNnl();
+   }
+   return(pc->SkipToMatch());
+}
+
+
+static bool for_needs_rewrite(Chunk *pc, E_Token desired_type)
+{
+   // The 'for' statement needs to be rewritten if `for(;;)` is not the
+   // preferred syntax for infinite loops and this 'for' is an infinite loop
+   // with no extra tokens (such as inline comments).
+
+   if (desired_type == CT_FOR)
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SPAREN_OPEN))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SEMICOLON))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SEMICOLON))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SPAREN_CLOSE))
+   {
+      return(false);
+   }
+   return(true);
+}
+
+
+static bool while_needs_rewrite(Chunk *keyword, E_Token desired_type, const char *desired_condition)
+{
+   // The 'while' statement needs to be rewritten if it has only the tokens that
+   // are strictly necessary (keyword, condition, two parentheses, and semicolon
+   // if do-while) and either the keyword or the condition needs to be changed.
+
+   Chunk *oparen    = keyword->GetNext();
+   Chunk *condition = oparen->GetNext();
+   Chunk *cparen    = condition->GetNext();
+
+   if (!chunk_is_token(oparen, CT_SPAREN_OPEN))
+   {
+      return(false);
+   }
+
+   if (!chunk_is_token(cparen, CT_SPAREN_CLOSE))
+   {
+      return(false);
+   }
+
+   if (chunk_is_token(keyword, CT_WHILE_OF_DO))
+   {
+      Chunk *semicolon = cparen->GetNext();
+
+      if (!chunk_is_token(semicolon, CT_SEMICOLON))
+      {
+         return(false);
+      }
+   }
+
+   if (!chunk_is_token(keyword, desired_type))
+   {
+      return(true);
+   }
```

Comment:
This seems wrong. If the result is true, it says it is ok to rewrite, without checking that the condition is "true" or "1".

Reply:
You're right; when changing from `while` to `for`, the function should still make sure that the original condition was `true` or `1`. I have now fixed that and added a test for it.

Reply:
great